# Подготовительный блок

Импортирую библиотеки:

In [ ]:
import pandas as pd
import numpy as np
import seaborn as sns
import squarify 
import matplotlib.pyplot as plt
import datetime
from datetime import timedelta
from datetime import datetime

Считываю датасеты:

In [ ]:
customers = pd.read_csv('olist_customers_dataset.csv') # покупатели
order_items = pd.read_csv('olist_order_items_dataset.csv', parse_dates = [4]) # товары в заказе
orders = pd.read_csv('olist_orders_dataset.csv', parse_dates=[3,4,5,6,7]) # заказы

Напишу функцию для разведывательного анализа данных, чтобы посмотреть основные сведения о датафреймах

In [ ]:
dataframes = {
    'customers': customers,
    'order_items': order_items,
    'orders': orders
}

def exp_data_analysis(df_name):
    print('-----------------------------------')
    print('Общая информация:')
    print('-----------------------------------')
    print('Имя датафрейма:', df_name)
    print('В датафрейме {} строк и {} столбцов.'.format(dataframes[df_name].shape[0], dataframes[df_name].shape[1]))
    print('-----------------------------------')
    print('Выведем типы данных:')
    print('-----------------------------------')
    print(dataframes[df_name].dtypes)
    print('-----------------------------------')
    print('Проверка на пропущенные значения:')
    print('-----------------------------------')
    print(dataframes[df_name].isna().sum())
    print('-----------------------------------')
    print('Проверка на дубликаты:')
    print('-----------------------------------')
    print(dataframes[df_name].loc[dataframes[df_name].duplicated()])
    print('-----------------------------------')
    print('Выведем первые 5 строк')
    print('-----------------------------------')

# Предварительный анализ данных
Посмотрим, какая информация хранится в наших датасетах, каков размер датасетов. <br>
Проверим, есть ли дубликаты и пропущенные значения.

## customers
Здесь хранятся уникальные идентификаторы пользователей

### Описание столбцов

+ customer_id              - позаказный идентификатор пользователя

+ customer_unique_id       - уникальный идентификатор пользователя  (аналог номера паспорта)

+ customer_zip_code_prefix - почтовый индекс пользователя

+ customer_city            - город доставки пользователя

+ customer_state           - штат доставки пользователя

In [ ]:
exp_data_analysis('customers')
customers.head()

-----------------------------------
Общая информация:
-----------------------------------
Имя датафрейма: customers
В датафрейме 99441 строк и 5 столбцов.
-----------------------------------
Выведем типы данных:
-----------------------------------
customer_id                 object
customer_unique_id          object
customer_zip_code_prefix     int64
customer_city               object
customer_state              object
dtype: object
-----------------------------------
Проверка на пропущенные значения:
-----------------------------------
customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64
-----------------------------------
Проверка на дубликаты:
-----------------------------------
Empty DataFrame
Columns: [customer_id, customer_unique_id, customer_zip_code_prefix, customer_city, customer_state]
Index: []
-----------------------------------
Выведем первые 5 строк
--------------

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

## order_items
Товарные позиции, входящие в заказы.

### Описание столбцов

+ order_id            —  уникальный идентификатор заказа (номер чека)

+ order_item_id       —  идентификатор товара внутри одного заказа

+ product_id          —  ид товара (аналог штрихкода)

+ seller_id           — ид производителя товара

+ shipping_limit_date —  максимальная дата доставки продавцом для передачи заказа партнеру по логистике

+ price               —  цена за единицу 

+ freight_value       —  вес товара

In [ ]:
exp_data_analysis('order_items')
customers.head()

-----------------------------------
Общая информация:
-----------------------------------
Имя датафрейма: order_items
В датафрейме 112650 строк и 7 столбцов.
-----------------------------------
Выведем типы данных:
-----------------------------------
order_id                       object
order_item_id                   int64
product_id                     object
seller_id                      object
shipping_limit_date    datetime64[ns]
price                         float64
freight_value                 float64
dtype: object
-----------------------------------
Проверка на пропущенные значения:
-----------------------------------
order_id               0
order_item_id          0
product_id             0
seller_id              0
shipping_limit_date    0
price                  0
freight_value          0
dtype: int64
-----------------------------------
Проверка на дубликаты:
-----------------------------------
Empty DataFrame
Columns: [order_id, order_item_id, product_id, seller_id, shippi

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

## orders
Данные о заказах

### Описание столбцов

+ order_id                      —  уникальный идентификатор заказа (номер чека)

+ customer_id                   —  позаказный идентификатор пользователя

+ order_status                  —  статус заказа

+ order_purchase_timestamp      —  время создания заказа

+ order_approved_at             —  время подтверждения оплаты заказа

+ order_delivered_carrier_date  —  время передачи заказа в логистическую службу

+ order_delivered_customer_date —  время доставки заказа

+ order_estimated_delivery_date —  обещанная дата доставки

### Статусы

+ created — создан

+ approved — подтверждён

+ invoiced — выставлен счёт

+ processing — в процессе сборки заказа

+ shipped — отгружен со склада

+ delivered — доставлен пользователю

+ unavailable — недоступен

+ canceled — отменён

In [ ]:
exp_data_analysis('orders')
customers.head()

-----------------------------------
Общая информация:
-----------------------------------
Имя датафрейма: orders
В датафрейме 99441 строк и 8 столбцов.
-----------------------------------
Выведем типы данных:
-----------------------------------
order_id                                 object
customer_id                              object
order_status                             object
order_purchase_timestamp         datetime64[ns]
order_approved_at                datetime64[ns]
order_delivered_carrier_date     datetime64[ns]
order_delivered_customer_date    datetime64[ns]
order_estimated_delivery_date    datetime64[ns]
dtype: object
-----------------------------------
Проверка на пропущенные значения:
-----------------------------------
order_id                            0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                 160
order_delivered_carrier_date     1783
order_delivered_custome

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

Пропущены некоторые даты, мои предположения такие:

*order_approved_at* - может быть пропущена, если оплата заказа еще не была произведена

*order_delivered_carrier_date* - может быть пропущена, если заказ еще не был передан курьеру

*order_delivered_customer_date* - может быть пропущена, если заказ еще не был доставлен покупателю

## Посмотрим для примера один заказ

Сначала выведем общую информацию о заказе:

In [ ]:
orders.query('order_id == "00143d0f86d6fbd9f9b38ab440ac16f5"')

order_id                       customer_id  \
36265  00143d0f86d6fbd9f9b38ab440ac16f5  2e45292979b9b2700ea09560eeb0f803   

      order_status order_purchase_timestamp   order_approved_at  \
36265    delivered      2017-10-16 15:29:43 2017-10-16 16:07:52   

      order_delivered_carrier_date order_delivered_customer_date  \
36265          2017-10-16 23:04:43           2017-10-27 18:43:56   

      order_estimated_delivery_date  
36265                    2017-11-06

Выведем информацию о товарных позициях, входящих в заказ:

In [ ]:
order_items.query('order_id == "00143d0f86d6fbd9f9b38ab440ac16f5"')

order_id  order_item_id  \
32  00143d0f86d6fbd9f9b38ab440ac16f5              1   
33  00143d0f86d6fbd9f9b38ab440ac16f5              2   
34  00143d0f86d6fbd9f9b38ab440ac16f5              3   

                          product_id                         seller_id  \
32  e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
33  e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   
34  e95ee6822b66ac6058e2e4aff656071a  a17f621c590ea0fab3d5d883e1630ec6   

   shipping_limit_date  price  freight_value  
32 2017-10-20 16:07:52  21.33           15.1  
33 2017-10-20 16:07:52  21.33           15.1  
34 2017-10-20 16:07:52  21.33           15.1

В этом заказе 3 одинаковых товара, учтем это. Для подсчета общей суммы заказа нам просто нужно взять сумму по столбцу *price*.

Наконец, выведем данные о заказчике:

In [ ]:
customers.query('customer_id == "2e45292979b9b2700ea09560eeb0f803"')

customer_id                customer_unique_id  \
58260  2e45292979b9b2700ea09560eeb0f803  bb8a37225e0279ac8a274c9765617eaf   

       customer_zip_code_prefix customer_city customer_state  
58260                     74663       goiania             GO

# Определение покупки

Покупкой буду считать такие заказы:

+ Факт оплаты подтвержден, то есть дата order_approved_at должна быть заполнена

+ Статус заказа - не unavailable (заказ недоступен, успешной покупкой считаться не может).

+ Статус заказа - не canceled (если заказ отменен, тоже считаться успешной покупкой не может).

+ Статус заказа - не created (потому что создание заказа еще не означает, что покупатель его оплатит, возможно он не подтвердит заказ или оформил заказ случайно).

In [ ]:
# формируем датасет с покупками
purchases = orders[orders.order_approved_at.notna() &
                                 (orders.order_status != "canceled") &
                                 (orders.order_status != "unavailable") &
                                 (orders.order_status != "created")]

In [ ]:
purchases.shape

(98188, 8)

Вот статусы наших заказов:

In [ ]:
purchases.value_counts(['order_status'])

order_status
delivered       96464
shipped          1107
invoiced          314
processing        301
approved            2
dtype: int64

Далее в заданиях будем использовать датасет purchases

# Задание 1. Сколько у нас пользователей, которые совершили покупку только один раз?



+ customer_id - позаказный идентификатор пользователя.

+ customer_unique_id - уникальный идентификатор пользователя (аналог номера паспорта).

Нам нужен именно *customer_unique_id*, потому что для каждого заказа генерируется свой *customer_id*, даже если по факту заказы сделал один и тот же человек.

Для этого нужно объединить данные из таблицы покупателей с таблицей покупок (данные с успешными покупками сформировали выше).

Нам необходим номер заказа, позаказный идентификатор пользователя и уникальный идентификатор пользователя.

In [ ]:
# объединяем датафреймы и оставляем только необходимые нам данные
the_only_purchase_customers = purchases \
                  .merge(customers, how = 'inner', on = 'customer_id') \
                  [['order_id', 'customer_id', 'customer_unique_id']] \
                  .groupby('customer_unique_id', as_index = False) \
                  .agg({'order_id':'nunique'}) \
                  .rename(columns={'order_id':'orders_count'}) \
                  .query('orders_count == 1') \
                  .agg({'customer_unique_id':'count'}) 

In [ ]:
the_only_purchase_customers

customer_unique_id    92087
dtype: int64

# Ответ
# Количество пользователей, совершивших только одну покупку - 92 087

- - - 

# Задание 2. Сколько заказов в месяц в среднем не доставляется по разным причинам (вывести детализацию по причинам)?

Для этого нам нужен будет исходный датасет (где есть данные по всем заказам, а не только по покупкам).

Я буду считать не доставленным тот заказ, у которого не заполнена дата *order_delivered_customer_date*.

Причин для этого может быть много. Например, заказ может быть еще у курьера, и просто не успел доехать до покупателя.

Но факт есть факт - заказ не доставлен, причины рассмотрю ниже.

In [ ]:
# собрали все недоставленные заказы
non_delivered_orders = orders[orders.order_delivered_customer_date.isna()]

Выведем статусы таких заказов:

In [ ]:
non_delivered_orders['order_status'].value_counts()

shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: order_status, dtype: int64

А теперь посмотрим на каждый статус в отдельности и проанализируем данные, нет ли в них ничего странного.

### shipped - отгружен со склада

In [ ]:
non_delivered_orders[(non_delivered_orders.order_status == "shipped")]

order_id                       customer_id  \
44     ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
154    6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   
162    36530871a5e80138db53bcfd8a104d90  4dafe3c841d2d6cc8a8b6d25b35704b9   
231    4d630f57194f5aba1a3d12ce23e71cd9  6d491c9fe2f04f6e2af6ec033cd8907c   
299    3b4ad687e7e5190db827e1ae5a8989dd  1a87b8517b7d31373b50396eb15cb445   
...                                 ...                               ...   
98534  a59ef0abffbef8ddaae23600b6ee6604  49e64271fb8eeb19851dc7e850e92204   
98779  dab8a6c6bd6ec448df5b3a6b6cb887bc  394653a10cab83cad40d7e2713f3ab89   
98932  492aed3c33bac22a8e04138319829283  58466c1166c377a56f6b2ae0d93ffbc0   
99113  274a7f7e4f1c17b7434a830e9b8759b1  670af30ca5b8c20878fecdafa5ee01b9   
99181  636cdd02667dc8d76d9296bf20a6890a  c162256b133c76f79181ce61d66545db   

      order_status order_purchase_timestamp   order_approved_at  \
44         shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
154        shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   
162        shipped      2017-05-09 11:48:37 2017-05-11 11:45:14   
231        shipped      2017-11-17 19:53:21 2017-11-18 19:50:31   
299        shipped      2018-06-28 12:52:15 2018-06-28 13:11:09   
...            ...                      ...                 ...   
98534      shipped      2017-11-06 17:35:49 2017-11-07 06:31:44   
98779      shipped      2018-07-14 10:12:51 2018-07-16 12:30:58   
98932      shipped      2018-07-06 16:26:47 2018-07-06 16:35:10   
99113      shipped      2018-06-23 13:25:15 2018-06-23 13:40:11   
99181      shipped      2018-02-17 14:31:22 2018-02-20 07:11:31   

      order_delivered_carrier_date order_delivered_customer_date  \
44             2018-06-05 14:32:00                           NaT   
154            2018-01-11 19:39:23                           NaT   
162            2017-05-11 13:21:47                           NaT   
231            2017-11-22 17:28:34                           NaT   
299            2018-07-04 15:20:00                           NaT   
...                            ...                           ...   
98534          2017-11-07 18:57:39                           NaT   
98779          2018-07-16 14:34:00                           NaT   
98932          2018-07-10 12:27:00                           NaT   
99113          2018-07-04 13:51:00                           NaT   
99181          2018-02-20 19:18:58                           NaT   

      order_estimated_delivery_date  
44                       2018-06-28  
154                      2018-02-07  
162                      2017-06-08  
231                      2017-12-13  
299                      2018-08-03  
...                             ...  
98534                    2017-11-24  
98779                    2018-08-03  
98932                    2018-07-26  
99113                    2018-07-24  
99181                    2018-03-14  

[1107 rows x 8 columns]

Похоже, здесь нет ничего необычного.

Эти заказы не доставлены просто потому, что они отгружены со склада и еще не доехали до покупателя.

### canceled - заказ отменен

Среди них есть заказы, у которых не заполнена ни дата подтверждения оплаты, ни дата передачи курьера.

Ничего подозрительного, такое вполне может быть, возможно пользователь сам отменил заказ после создания.

In [ ]:
non_delivered_orders[non_delivered_orders.order_delivered_carrier_date.isna() &
                     non_delivered_orders.order_approved_at.isna() &
                    (non_delivered_orders.order_status == "canceled")]

order_id                       customer_id  \
1130   00b1cb0320190ca0daa2c88b35206009  3532ba38a3fd242259a514ac2b6ae6b6   
1801   ed3efbd3a87bea76c2812c66a0b32219  191984a8ba4cbb2145acb4fe35b69664   
1868   df8282afe61008dc26c6c31011474d02  aa797b187b5466bc6925aaaa4bb3bed1   
2029   8d4c637f1accf7a88a4555f02741e606  b1dd715db389a2077f43174e7a675d07   
2161   7a9d4c7f9b068337875b95465330f2fc  7f71ae48074c0cfec9195f88fcbfac55   
...                                 ...                               ...   
97696  5a00b4d35edffc56b825c3646a99ba9d  6a3bdf004ca96338fb5fad1b8d93c2e6   
98415  227c804e2a44760671a6a5697ea549e4  62e7477e75e542243ee62a0ba73f410f   
98909  e49e7ce1471b4693482d40c2bd3ad196  e4e7ab3f449aeb401f0216f86c2104db   
99283  3a3cddda5a7c27851bd96c3313412840  0b0d6095c5555fe083844281f6b093bb   
99347  a89abace0dcc01eeb267a9660b5ac126  2f0524a7b1b3845a1a57fcf3910c4333   

      order_status order_purchase_timestamp order_approved_at  \
1130      canceled      2018-08-28 15:26:39               NaT   
1801      canceled      2018-09-20 13:54:16               NaT   
1868      canceled      2017-03-04 12:14:30               NaT   
2029      canceled      2018-08-29 16:27:49               NaT   
2161      canceled      2017-05-01 16:12:39               NaT   
...            ...                      ...               ...   
97696     canceled      2017-07-02 15:38:46               NaT   
98415     canceled      2017-09-28 15:02:56               NaT   
98909     canceled      2018-08-07 11:16:28               NaT   
99283     canceled      2018-08-31 16:13:44               NaT   
99347     canceled      2018-09-06 18:45:47               NaT   

      order_delivered_carrier_date order_delivered_customer_date  \
1130                           NaT                           NaT   
1801                           NaT                           NaT   
1868                           NaT                           NaT   
2029                           NaT                           NaT   
2161                           NaT                           NaT   
...                            ...                           ...   
97696                          NaT                           NaT   
98415                          NaT                           NaT   
98909                          NaT                           NaT   
99283                          NaT                           NaT   
99347                          NaT                           NaT   

      order_estimated_delivery_date  
1130                     2018-09-12  
1801                     2018-10-17  
1868                     2017-04-10  
2029                     2018-09-13  
2161                     2017-05-30  
...                             ...  
97696                    2017-07-25  
98415                    2017-10-16  
98909                    2018-08-10  
99283                    2018-10-01  
99347                    2018-09-27  

[141 rows x 8 columns]

А вот у этих заказов дата оплаты проставлена, но заказ все равно не был передан курьеру и был отменен.

Такое тоже может быть.

In [ ]:
non_delivered_orders[non_delivered_orders.order_delivered_carrier_date.isna() &
                     non_delivered_orders.order_approved_at.notna() &
                    (non_delivered_orders.order_status == "canceled")]

order_id                       customer_id  \
397    1b9ecfe83cdc259250e1a8aca174f0ad  6d6b50b66d79f80827b6d96751528d30   
1811   0966b61e30c4a07edbd7523f59b3f3e4  2fcc597b8934d99715dbfff7909dd27f   
1819   9021cf1919f615a121410790dcce848f  7acf55df0298e1d2c31200fb4f6fb93b   
1971   a39d3db795a5cf4c8b6c9dd050f0d326  ec66df2cb66dfda07c03050470e21f69   
2197   375f80ffc05feb35ce2680f1fa490575  8f6e5911262a8742393596bee2c09eac   
...                                 ...                               ...   
98496  2be6c8b958e16856397259974d4abee9  ccbce7bcb13ee88adcf78890cae0ea42   
98629  f5f8998eee8ec7bc513dc52847d64ce0  f4656b824844a039a87fd9c51ad3586a   
98771  5a8a4dc28b16fb90469ad749f9535773  c0c8b8bb055100a0cc08dcc04d847ac9   
98791  b159d0ce7cd881052da94fa165617b05  e0c3bc5ce0836b975d6b2a8ce7bb0e3e   
99143  6560fb10610771449cb0463c5ba12199  0d07d0a588caf93cc66b7a8aff86d2fe   

      order_status order_purchase_timestamp   order_approved_at  \
397       canceled      2018-08-04 14:29:27 2018-08-07 04:10:26   
1811      canceled      2018-05-22 18:50:55 2018-05-22 19:17:15   
1819      canceled      2018-07-04 16:05:56 2018-07-06 02:55:16   
1971      canceled      2017-03-13 16:12:24 2017-03-13 16:12:24   
2197      canceled      2017-06-13 21:10:05 2017-06-13 21:22:35   
...            ...                      ...                 ...   
98496     canceled      2018-04-02 18:20:00 2018-04-02 18:30:38   
98629     canceled      2018-03-01 11:42:23 2018-03-01 12:20:32   
98771     canceled      2018-03-13 10:58:09 2018-03-14 03:08:35   
98791     canceled      2017-03-11 19:51:36 2017-03-11 19:51:36   
99143     canceled      2017-10-01 22:26:25 2017-10-01 22:35:22   

      order_delivered_carrier_date order_delivered_customer_date  \
397                            NaT                           NaT   
1811                           NaT                           NaT   
1819                           NaT                           NaT   
1971                           NaT                           NaT   
2197                           NaT                           NaT   
...                            ...                           ...   
98496                          NaT                           NaT   
98629                          NaT                           NaT   
98771                          NaT                           NaT   
98791                          NaT                           NaT   
99143                          NaT                           NaT   

      order_estimated_delivery_date  
397                      2018-08-14  
1811                     2018-07-03  
1819                     2018-08-22  
1971                     2017-03-30  
2197                     2017-07-06  
...                             ...  
98496                    2018-04-19  
98629                    2018-03-20  
98771                    2018-03-23  
98791                    2017-03-30  
99143                    2017-10-27  

[409 rows x 8 columns]

А еще есть заказы у которых подтверждена оплата, они переданы курьеру, но до покупателя все равно не доехали.

В принципе, тоже реально. Например, у покупателя случился форс-мажор, он понимает, что не сможет встретить курьера с доставкой, и отменяет заказ.

In [ ]:
non_delivered_orders[non_delivered_orders.order_delivered_carrier_date.notna() &
                     non_delivered_orders.order_approved_at.notna() &
                    (non_delivered_orders.order_status == "canceled")]

order_id                       customer_id  \
613    714fb133a6730ab81fa1d3c1b2007291  e3fe72696c4713d64d3c10afe71e75ed   
1058   3a129877493c8189c59c60eb71d97c29  0913cdce793684e52bbfac69d87e91fd   
5127   8a9cce19aa49e055f2babb81d5a71c40  f2676ef361959c938b35ecd32321adce   
5449   c1049eef89915c6beb3604bf3d6588d7  ed16865bcadcea1a00a928bf418a74bf   
5509   885935156ba43094f3ae95f3c14e0381  3885c59916b2b261205c303fe0a10abf   
...                                 ...                               ...   
93172  47e96b5cb6f0592d336b6bf8a817d1c3  9cce7727e7d1d1a738ccc755eae38e93   
94175  c9f06c1229ec7cd4ec9aed55d434b499  fa89053140d09081b3a3a4608a74c083   
96300  1def7af24cb7182497acfd953f6dd88b  b35c38c2cd6a72b58cb104154d9782e4   
97207  716bcd6dfe99970baf4aed2083136ed8  42715bff8f22b8dc62e5055b1e9fef7e   
98397  affcfd3067c168967157abe801a3c4ec  5f40f0d1ada169f5b00ec600e563826a   

      order_status order_purchase_timestamp   order_approved_at  \
613       canceled      2018-01-26 21:34:08 2018-01-26 21:58:39   
1058      canceled      2018-01-25 13:34:24 2018-01-25 13:50:20   
5127      canceled      2018-02-01 10:04:20 2018-02-03 02:52:42   
5449      canceled      2018-02-05 23:08:11 2018-02-05 23:29:26   
5509      canceled      2018-02-02 16:06:44 2018-02-06 04:32:39   
...            ...                      ...                 ...   
93172     canceled      2018-01-21 12:44:16 2018-01-22 13:49:10   
94175     canceled      2018-02-05 20:26:29 2018-02-05 20:49:48   
96300     canceled      2017-12-21 09:22:36 2017-12-23 02:12:37   
97207     canceled      2018-02-03 13:37:35 2018-02-03 13:50:26   
98397     canceled      2018-02-17 22:18:30 2018-02-17 22:35:29   

      order_delivered_carrier_date order_delivered_customer_date  \
613            2018-01-29 22:33:25                           NaT   
1058           2018-01-26 21:42:18                           NaT   
5127           2018-02-05 20:41:28                           NaT   
5449           2018-02-09 22:13:01                           NaT   
5509           2018-02-07 20:48:31                           NaT   
...                            ...                           ...   
93172          2018-02-01 00:08:56                           NaT   
94175          2018-02-06 19:49:32                           NaT   
96300          2018-01-30 19:28:55                           NaT   
97207          2018-02-05 19:59:59                           NaT   
98397          2018-02-19 20:08:21                           NaT   

      order_estimated_delivery_date  
613                      2018-02-22  
1058                     2018-02-23  
5127                     2018-02-23  
5449                     2018-03-07  
5509                     2018-03-08  
...                             ...  
93172                    2018-02-19  
94175                    2018-03-01  
96300                    2018-02-20  
97207                    2018-03-01  
98397                    2018-03-06  

[69 rows x 8 columns]

В сумме как раз получается 619 заказов, ничего необычного нет, с этим статусом разобрались.

Заказы были отменены, поэтому они не были доставлены.

### unavailable - заказ недоступен

In [ ]:
non_delivered_orders[non_delivered_orders.order_delivered_carrier_date.isna() &
                     (non_delivered_orders.order_status == "unavailable")]

order_id                       customer_id  \
266    8e24261a7e58791d10cb1bf9da94df5c  64a254d30eed42cd0e6c36dddb88adf0   
586    c272bcd21c287498b4883c7512019702  9582c5bbecc65eb568e2c1d839b5cba1   
687    37553832a3a89c9b2db59701c357ca67  7607cd563696c27ede287e515812d528   
737    d57e15fb07fd180f06ab3926b39edcd2  470b93b3f1cde85550fc74cd3a476c78   
1160   2f634e2cebf8c0283e7ef0989f77d217  7353b0fb8e8d9675e3a704c60ca44ebe   
...                                 ...                               ...   
99147  6cf5c427fb11bc5fe8abefa86aa7ba0b  91ac14a1e50bc022eeecf0c2ba37006a   
99233  1c7c8ab324927f5b6e52338bd110bef4  95b1767df51d6fa09402267d8de314df   
99252  aaab15da689073f8f9aa978a390a69d1  df20748206e4b865b2f14a5eabbfcf34   
99348  a69ba794cc7deb415c3e15a0a3877e69  726f0894b5becdf952ea537d5266e543   
99415  5fabc81b6322c8443648e1b21a6fef21  32c9df889d41b0ee8309a5efb6855dcb   

      order_status order_purchase_timestamp   order_approved_at  \
266    unavailable      2017-11-16 15:09:28 2017-11-16 15:26:57   
586    unavailable      2018-01-31 11:31:37 2018-01-31 14:23:50   
687    unavailable      2017-08-14 17:38:02 2017-08-17 00:15:18   
737    unavailable      2018-01-08 19:39:03 2018-01-09 07:26:08   
1160   unavailable      2017-09-27 20:55:33 2017-09-28 01:32:50   
...            ...                      ...                 ...   
99147  unavailable      2017-10-17 10:53:48 2017-10-19 03:12:43   
99233  unavailable      2017-12-23 18:48:09 2017-12-23 18:54:22   
99252  unavailable      2018-01-16 14:27:59 2018-01-17 03:37:34   
99348  unavailable      2017-08-23 16:28:04 2017-08-28 15:44:47   
99415  unavailable      2017-10-10 10:50:03 2017-10-14 18:35:57   

      order_delivered_carrier_date order_delivered_customer_date  \
266                            NaT                           NaT   
586                            NaT                           NaT   
687                            NaT                           NaT   
737                            NaT                           NaT   
1160                           NaT                           NaT   
...                            ...                           ...   
99147                          NaT                           NaT   
99233                          NaT                           NaT   
99252                          NaT                           NaT   
99348                          NaT                           NaT   
99415                          NaT                           NaT   

      order_estimated_delivery_date  
266                      2017-12-05  
586                      2018-02-16  
687                      2017-09-05  
737                      2018-02-06  
1160                     2017-10-27  
...                             ...  
99147                    2017-10-27  
99233                    2018-01-19  
99252                    2018-02-06  
99348                    2017-09-15  
99415                    2017-10-23  

[609 rows x 8 columns]

Похоже, с этими заказами что-то случилось, дата подтверждения оплаты указана, но заказ не уехал со склада, не был передан курьеру и не доехал до покупателя. Возможно, была какая-то ошибка в базе данных, и покупатели заказали тот товар, которого по факту уже не было в наличии

### invoiced - выставлен счет

In [ ]:
non_delivered_orders[non_delivered_orders.order_delivered_carrier_date.isna() &
                     (non_delivered_orders.order_status == "invoiced")]

order_id                       customer_id  \
6      136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
103    0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
455    38b7efdf33dd5561f4f5d4f6e07b0414  021e84751ba0ead75b6d314a6ead88d9   
1833   51b0dccc8596ce37a930dff2d63a10a2  31bf1057b00f14804278590bbac18b1b   
2116   5504eaa5a86eb25fa666cf2e6b96c701  a30be38e1ed0ffb39b318bf1d15e98d9   
...                                 ...                               ...   
98414  41ad2ae32ee66c07ef3b0301ba02ad27  5c7f8d14a0a12198d1d07167d7d343b3   
98624  9ffb325177b96c422447e3b9ee9a6270  e4079117ea0dc8bc27274e34c4796515   
98845  021dac6e3120485db6a77474aa21b979  c40f7f00e91eb8a0da1e58484380a796   
99031  c526fe83c01c48066394ff5d81f9ab9a  6c678c38f3a0874b7bfeeb2d5fb22876   
99066  46bdec7cc92d926500bee756ee1df6dc  0cfcefc0ed0af6a2e601540feeac24e7   

      order_status order_purchase_timestamp   order_approved_at  \
6         invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
103       invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
455       invoiced      2017-08-01 18:17:41 2017-08-01 18:32:30   
1833      invoiced      2017-05-05 22:34:48 2017-05-05 22:45:12   
2116      invoiced      2017-11-29 08:33:06 2017-11-29 08:56:23   
...            ...                      ...                 ...   
98414     invoiced      2017-11-24 15:54:47 2017-11-24 19:33:24   
98624     invoiced      2018-05-10 19:11:18 2018-05-10 19:37:27   
98845     invoiced      2017-11-24 22:46:09 2017-11-25 01:53:42   
99031     invoiced      2018-08-07 11:11:30 2018-08-07 11:25:01   
99066     invoiced      2018-07-16 13:20:06 2018-07-16 13:30:35   

      order_delivered_carrier_date order_delivered_customer_date  \
6                              NaT                           NaT   
103                            NaT                           NaT   
455                            NaT                           NaT   
1833                           NaT                           NaT   
2116                           NaT                           NaT   
...                            ...                           ...   
98414                          NaT                           NaT   
98624                          NaT                           NaT   
98845                          NaT                           NaT   
99031                          NaT                           NaT   
99066                          NaT                           NaT   

      order_estimated_delivery_date  
6                        2017-05-09  
103                      2018-08-21  
455                      2017-08-28  
1833                     2017-06-06  
2116                     2017-12-20  
...                             ...  
98414                    2017-12-18  
98624                    2018-06-05  
98845                    2017-12-15  
99031                    2018-08-17  
99066                    2018-08-06  

[314 rows x 8 columns]

Похоже на то, что это та стадия заказа, когда покупателю выставлен счет (на это указывает статус invoiced), оплата подтверждена (на это указывает заполненное поле order_approved_at), но заказ пока ожидает сборки.

### processing - в процессе сборки заказа

In [ ]:
non_delivered_orders.query('order_status == "processing"')

order_id                       customer_id  \
128    15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
324    d3c8851a6651eeff2f73b0e011ac45d0  957f8e082185574de25992dc659ebbc0   
741    6a6c7d523fd59eb5bbefc007331af717  d954782ec6c0e911292c8a80757ef28d   
1192   745e1d4a7f8c4b548881788d4113bb1d  7198d7088442e4ddfe553353d8ddc957   
1516   1d52ba7197c7acebbb4f826f6585536f  c9c7fe860d602373a9e93f8bfe9d877a   
...                                 ...                               ...   
97400  dcdfc540e42725663242bb884c28f0a6  38972104038aa68fcc61277dbf6e7ca9   
97666  e471815e7114cdb474064f7dbb1a8b67  092c9316ae71b2fe43e526043f351967   
98089  10951d02d64917a34959abeb8130601e  3e6754e591ff3568ccc5bf69a649918c   
99140  aea0db338150b526dde24f6fd953a5ed  379a02efdc6a56bd27f99b95fc2f6c06   
99313  e9e64a17afa9653aacf2616d94c005b8  b4cd0522e632e481f8eaf766a2646e86   

      order_status order_purchase_timestamp   order_approved_at  \
128     processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
324     processing      2016-10-05 22:44:13 2016-10-06 15:51:05   
741     processing      2017-11-24 20:09:33 2017-11-24 23:15:15   
1192    processing      2017-11-17 06:34:36 2017-11-18 02:15:40   
1516    processing      2017-02-13 18:32:55 2017-02-13 18:43:55   
...            ...                      ...                 ...   
97400   processing      2017-10-30 10:46:44 2017-10-30 11:09:55   
97666   processing      2017-12-20 11:00:02 2017-12-20 11:10:43   
98089   processing      2018-02-13 21:02:02 2018-02-15 04:11:21   
99140   processing      2017-12-26 21:56:13 2017-12-26 22:05:26   
99313   processing      2018-01-05 23:07:24 2018-01-09 07:18:05   

      order_delivered_carrier_date order_delivered_customer_date  \
128                            NaT                           NaT   
324                            NaT                           NaT   
741                            NaT                           NaT   
1192                           NaT                           NaT   
1516                           NaT                           NaT   
...                            ...                           ...   
97400                          NaT                           NaT   
97666                          NaT                           NaT   
98089                          NaT                           NaT   
99140                          NaT                           NaT   
99313                          NaT                           NaT   

      order_estimated_delivery_date  
128                      2017-10-03  
324                      2016-12-09  
741                      2017-12-20  
1192                     2017-12-12  
1516                     2017-04-04  
...                             ...  
97400                    2017-11-23  
97666                    2018-02-02  
98089                    2018-03-13  
99140                    2018-01-26  
99313                    2018-02-06  

[301 rows x 8 columns]

Ничего необычного, похоже что на этом этапе заказ собирается.

### delivered - доставлен пользователю

Интересно, как сюда попал этот заказ. Мы ведь ограничивали данные по столбцу order_delivered_customer_date.

In [ ]:
non_delivered_orders.query('order_status == "delivered"')

order_id                       customer_id  \
3002   2d1e2d5bf4dc7227b3bfebb81328c15f  ec05a6d8558c6455f0cbbd8a420ad34f   
20618  f5dd62b788049ad9fc0526e3ad11a097  5e89028e024b381dc84a13a3570decb4   
43834  2ebdfc4f15f23b91474edf87475f108e  29f0540231702fda0cfdee0a310f11aa   
79263  e69f75a717d64fc5ecdfae42b2e8e086  cfda40ca8dd0a5d486a9635b611b398a   
82868  0d3268bad9b086af767785e3f0fc0133  4f1d63d35fb7c8999853b2699f5c7649   
92643  2d858f451373b04fb5c984a1cc2defaf  e08caf668d499a6d643dafd7c5cc498a   
97647  ab7c89dc1bf4a1ead9d6ec1ec8968a84  dd1b84a7286eb4524d52af4256c0ba24   
98038  20edc82cf5400ce95e1afacc25798b31  28c37425f1127d887d7337f284080a0f   

      order_status order_purchase_timestamp   order_approved_at  \
3002     delivered      2017-11-28 17:44:07 2017-11-28 17:56:40   
20618    delivered      2018-06-20 06:58:43 2018-06-20 07:19:05   
43834    delivered      2018-07-01 17:05:11 2018-07-01 17:15:12   
79263    delivered      2018-07-01 22:05:55 2018-07-01 22:15:14   
82868    delivered      2018-07-01 21:14:02 2018-07-01 21:29:54   
92643    delivered      2017-05-25 23:22:43 2017-05-25 23:30:16   
97647    delivered      2018-06-08 12:09:39 2018-06-08 12:36:39   
98038    delivered      2018-06-27 16:09:12 2018-06-27 16:29:30   

      order_delivered_carrier_date order_delivered_customer_date  \
3002           2017-11-30 18:12:23                           NaT   
20618          2018-06-25 08:05:00                           NaT   
43834          2018-07-03 13:57:00                           NaT   
79263          2018-07-03 13:57:00                           NaT   
82868          2018-07-03 09:28:00                           NaT   
92643                          NaT                           NaT   
97647          2018-06-12 14:10:00                           NaT   
98038          2018-07-03 19:26:00                           NaT   

      order_estimated_delivery_date  
3002                     2017-12-18  
20618                    2018-07-16  
43834                    2018-07-30  
79263                    2018-07-30  
82868                    2018-07-24  
92643                    2017-06-23  
97647                    2018-06-26  
98038                    2018-07-19

Есть один заказ, у которого не указана дата передачи курьеру, и также не указана дата доставки покупателю.
И еще 7 заказов, которые по статусу доставлены, а дата доставки получателю - пустая.

Я бы взяла на контроль такие заказы, возможно это техническая ошибка, а возможно тут кроются какие-то другие причины, в любом случае их немного и можно разобраться с каждым.

### created - заказ создан

In [ ]:
non_delivered_orders.query('order_status == "created"')

order_id                       customer_id  \
7434   b5359909123fa03c50bdb0cfed07f098  438449d4af8980d107bf04571413a8e7   
9238   dba5062fbda3af4fb6c33b1e040ca38f  964a6df3d9bdf60fe3e7b8bb69ed893a   
21441  7a4df5d8cff4090e541401a20a22bb80  725e9c75605414b21fd8c8d5a1c2f1d6   
55086  35de4050331c6c644cddc86f4f2d0d64  4ee64f4bfc542546f422da0aeb462853   
58958  90ab3e7d52544ec7bc3363c82689965f  7d61b9f4f216052ba664f22e9c504ef1   

      order_status order_purchase_timestamp order_approved_at  \
7434       created      2017-12-05 01:07:52               NaT   
9238       created      2018-02-09 17:21:04               NaT   
21441      created      2017-11-25 11:10:33               NaT   
55086      created      2017-12-05 01:07:58               NaT   
58958      created      2017-11-06 13:12:34               NaT   

      order_delivered_carrier_date order_delivered_customer_date  \
7434                           NaT                           NaT   
9238                           NaT                           NaT   
21441                          NaT                           NaT   
55086                          NaT                           NaT   
58958                          NaT                           NaT   

      order_estimated_delivery_date  
7434                     2018-01-11  
9238                     2018-03-07  
21441                    2017-12-12  
55086                    2018-01-08  
58958                    2017-12-01

Эти заказы только-только созданы, ничего криминального.

### approved - заказ подтвержден

In [ ]:
non_delivered_orders.query('order_status == "approved"')

order_id                       customer_id  \
44897  a2e4c44360b4a57bdff22f3a4630c173  8886130db0ea6e9e70ba0b03d7c0d286   
88457  132f1e724165a07f6362532bfb97486e  b2191912d8ad6eac2e4dc3b6e1459515   

      order_status order_purchase_timestamp   order_approved_at  \
44897     approved      2017-02-06 20:18:17 2017-02-06 20:30:19   
88457     approved      2017-04-25 01:25:34 2017-04-30 20:32:41   

      order_delivered_carrier_date order_delivered_customer_date  \
44897                          NaT                           NaT   
88457                          NaT                           NaT   

      order_estimated_delivery_date  
44897                    2017-03-01  
88457                    2017-05-22

Как вариант, эти заказы приняты, подтверждены, но счет еще не отправлен покупателю.

Создадим функцию для определения причины:

In [ ]:
#define function for classifying orders based on statuses

def define_reason(row):
 if row['order_status'] == 'shipped':
     val = 'Товар отгружен со склада'
 elif row['order_status'] == 'canceled':
     val = 'Заказ отменен'
 elif row['order_status'] == 'unavailable':
     val = 'Заказ недоступен'
 elif row['order_status'] == 'invoiced':
     val = 'Заказ еще не оплачен, выставлен счет'
 elif row['order_status'] == 'processing':
     val = 'Заказ еще в процессе сборки'
 elif row['order_status'] == 'delivered':
     val = 'Возможна ошибка, ПРОВЕРИТЬ ЭТИ ЗАКАЗЫ'
 elif row['order_status'] == 'created':
     val = 'Заказ в статусе создан'
 elif row['order_status'] == 'approved':
     val = 'Заказ подтвержден, еще не оплачен'
 return val

Теперь создаем столбец, где будет указана причина того, что заказ пока не доставлен

In [ ]:
#create new column 'non_delivered_reason' using the function above
non_delivered_orders['non_delivered_reason'] = non_delivered_orders.apply (define_reason, axis=1)

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


Посмотрим на результат, столбец добавился, можно анализировать дальше.

In [ ]:
non_delivered_orders.head()

order_id                       customer_id  \
6    136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44   ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103  0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128  15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154  6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   

    order_status order_purchase_timestamp   order_approved_at  \
6       invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
44       shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
103     invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128   processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
154      shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   

    order_delivered_carrier_date order_delivered_customer_date  \
6                            NaT                           NaT   
44           2018-06-05 14:32:00                           NaT   
103                          NaT                           NaT   
128                          NaT                           NaT   
154          2018-01-11 19:39:23                           NaT   

    order_estimated_delivery_date                  non_delivered_reason  
6                      2017-05-09  Заказ еще не оплачен, выставлен счет  
44                     2018-06-28              Товар отгружен со склада  
103                    2018-08-21  Заказ еще не оплачен, выставлен счет  
128                    2017-10-03           Заказ еще в процессе сборки  
154                    2018-02-07              Товар отгружен со склада

При подсчете количества недоставленных товаров в месяц будем ориентироваться на столбец *order_purchase_timestamp*

Создадим дополнительный столбец, который будет содержать эту дату, обрезанную до начала месяца.

In [ ]:
non_delivered_orders['order_purchase_month'] = non_delivered_orders['order_purchase_timestamp'].dt.to_period("M")

/opt/tljh/user/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


Проверим, как выглядит новый столбец. Все получилось, можно идти дальше.

In [ ]:
non_delivered_orders.head()

order_id                       customer_id  \
6    136cce7faa42fdb2cefd53fdc79a6098  ed0271e0b7da060a393796590e7b737a   
44   ee64d42b8cf066f35eac1cf57de1aa85  caded193e8e47b8362864762a83db3c5   
103  0760a852e4e9d89eb77bf631eaaf1c84  d2a79636084590b7465af8ab374a8cf5   
128  15bed8e2fec7fdbadb186b57c46c92f2  f3f0e613e0bdb9c7cee75504f0f90679   
154  6942b8da583c2f9957e990d028607019  52006a9383bf149a4fb24226b173106f   

    order_status order_purchase_timestamp   order_approved_at  \
6       invoiced      2017-04-11 12:22:08 2017-04-13 13:25:17   
44       shipped      2018-06-04 16:44:48 2018-06-05 04:31:18   
103     invoiced      2018-08-03 17:44:42 2018-08-07 06:15:14   
128   processing      2017-09-03 14:22:03 2017-09-03 14:30:09   
154      shipped      2018-01-10 11:33:07 2018-01-11 02:32:30   

    order_delivered_carrier_date order_delivered_customer_date  \
6                            NaT                           NaT   
44           2018-06-05 14:32:00                           NaT   
103                          NaT                           NaT   
128                          NaT                           NaT   
154          2018-01-11 19:39:23                           NaT   

    order_estimated_delivery_date                  non_delivered_reason  \
6                      2017-05-09  Заказ еще не оплачен, выставлен счет   
44                     2018-06-28              Товар отгружен со склада   
103                    2018-08-21  Заказ еще не оплачен, выставлен счет   
128                    2017-10-03           Заказ еще в процессе сборки   
154                    2018-02-07              Товар отгружен со склада   

    order_purchase_month  
6                2017-04  
44               2018-06  
103              2018-08  
128              2017-09  
154              2018-01

In [ ]:
non_delivered_orders_stat = non_delivered_orders.groupby(by=['non_delivered_reason','order_purchase_month'], as_index=False) \
                    .agg({'order_id':'count'}) \
                    .rename(columns={"order_id": "orders_count"}) \
                    .groupby(by=['non_delivered_reason'], as_index=False) \
                    .agg({'orders_count':'mean'}) \
                    .rename(columns={"orders_count": "mean_cancelled_orders_count"})

In [ ]:
non_delivered_orders_stat.sort_values('mean_cancelled_orders_count', ascending = False)

non_delivered_reason  mean_cancelled_orders_count
7               Товар отгружен со склада                    48.130435
4                       Заказ недоступен                    29.000000
5                          Заказ отменен                    25.791667
2            Заказ еще в процессе сборки                    15.842105
3   Заказ еще не оплачен, выставлен счет                    14.952381
0  Возможна ошибка, ПРОВЕРИТЬ ЭТИ ЗАКАЗЫ                     2.000000
1                 Заказ в статусе создан                     1.666667
6      Заказ подтвержден, еще не оплачен                     1.000000

- - -

# 3. По каждому товару определить, в какой день недели товар чаще всего покупается

Здесь уже будем использовать датасет с покупками - *purchases*.

Вспомним, какие данные там есть:

In [ ]:
purchases.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

Есть информация о заказах, но нет товаров, которые были в конкретном заказе. Значит, нужно взять еще датасет *order_items* - товарные позиции, входящие в заказы.

Не будем забывать, что при соединении датасетов у нас размножатся строки, потому что одному заказу может соответствовать несколько строк из списка товарных позиций.

Из датасета purchases нам нужны столбцы *order_id, order_purchase_timestamp*

Из датасета *olist_order_items_dataset* нам нужны столбцы *order_id, order_item_id, product_id*

In [ ]:
# соберем нужные данные
purchases2items = purchases.merge(order_items, how = 'inner', on = 'order_id') \
                            [['order_id', 'order_purchase_timestamp', 'order_item_id', 'product_id']]

In [ ]:
# добавим столбец с названием дня недели
purchases2items['order_purchase_weekday'] = purchases2items['order_purchase_timestamp'].dt.day_name()

In [ ]:
# наконец, соберем статистику по дням
products2weekdays = purchases2items.groupby(['product_id','order_purchase_weekday'], as_index=False) \
                                 .agg({'order_id':'count'}) \
                                 .rename(columns={"order_id": "purchases_count"}) \
                                 .sort_values(by="purchases_count", ascending=False) \
                                 .drop_duplicates(subset='product_id')

In [ ]:
products2weekdays.head()

product_id order_purchase_weekday  \
15812  422879e10f46682990de24d770e7f83d              Wednesday   
36305  99a4788cb24856965c36a24e339b6058                 Monday   
40819  aca2eb7d00ea1a7b8ebd4e68314663af               Thursday   
20009  53b36df67ebb7c41585e8d54d6772e08                Tuesday   
13459  389d119b48cf3043d311335e499d9c6b               Thursday   

       purchases_count  
15812               93  
36305               92  
40819               89  
20009               76  
13459               67

- - -

# 4. Сколько у каждого из пользователей в среднем покупок в неделю (по месяцам)? Не стоит забывать, что внутри месяца может быть не целое количество недель. Например, в ноябре 2021 года 4,28 недели. И внутри метрики это нужно учесть.

Нам нужен датасет *purchases* + датасет с данными пользователей (оттуда нужен id пользователя уникальный, а не позаказный).

Из датасета по пользователям нам нужен *customer_unique_id*.

Из датасета *purchases* нам нужен столбец *order_id, customer_id и order_purchase_timestamp*

In [ ]:
customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [ ]:
purchases.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

In [ ]:
purch2items = purchases.merge(customers, how = 'inner', on = 'customer_id') \
                        [['order_id','order_purchase_timestamp','customer_unique_id']]

Для дальнейших рассчетов нам понадобится знать год и номер месяца, для расчета количества дней

In [ ]:
purch2items['order_purchase_year'] = purch2items['order_purchase_timestamp'].dt.year
purch2items['order_purchase_month'] = purch2items['order_purchase_timestamp'].dt.month
purch2items['order_purchase_day'] = purch2items['order_purchase_timestamp'].dt.day

In [ ]:
purch2items.head()

order_id order_purchase_timestamp  \
0  e481f51cbdc54678b7cc49136f2d6af7      2017-10-02 10:56:33   
1  53cdb2fc8bc7dce0b6741e2150273451      2018-07-24 20:41:37   
2  47770eb9100c2d0c44946d9cf07ec65d      2018-08-08 08:38:49   
3  949d5b44dbf5de918fe9c16f97b45f8a      2017-11-18 19:28:06   
4  ad21c59c0840e6cb83a9ceb5573f8159      2018-02-13 21:18:39   

                 customer_unique_id  order_purchase_year  \
0  7c396fd4830fd04220f754e42b4e5bff                 2017   
1  af07308b275d755c9edb36a90c618231                 2018   
2  3a653a41f6f9fc3d2a113cf8398680e8                 2018   
3  7c142cf63193a1473d2e66489a9ae977                 2017   
4  72632f0f9dd73dfee390c9b22eb56dd6                 2018   

   order_purchase_month  order_purchase_day  
0                    10                   2  
1                     7                  24  
2                     8                   8  
3                    11                  18  
4                     2                  13

Проверю, целиком ли учтены месяцы в наших датасетах, нет ли такого, что в каком-то месяце не хватает дней.

In [ ]:
purch2items.groupby(by=['order_purchase_year','order_purchase_month'], as_index = False) \
           .agg({'order_purchase_day':'nunique'}) \
           .sort_values(by=["order_purchase_year","order_purchase_month"])

order_purchase_year  order_purchase_month  order_purchase_day
0                  2016                     9                   2
1                  2016                    10                   8
2                  2016                    12                   1
3                  2017                     1                  27
4                  2017                     2                  28
5                  2017                     3                  31
6                  2017                     4                  30
7                  2017                     5                  31
8                  2017                     6                  30
9                  2017                     7                  31
10                 2017                     8                  31
11                 2017                     9                  30
12                 2017                    10                  31
13                 2017                    11                  30
14                 2017                    12                  31
15                 2018                     1                  31
16                 2018                     2                  28
17                 2018                     3                  31
18                 2018                     4                  30
19                 2018                     5                  31
20                 2018                     6                  30
21                 2018                     7                  31
22                 2018                     8                  29
23                 2018                     9                   1

Итак, видно, что в основном месяцы полные.

Однако, сентябрь, октябрь, декабрь 2016 года заполнено лишь частично, данных за ноябрь 2016 и вовсе нет.
Январь 2017 - не полный месяц, всего 27 дней из 31.

Август 2018 года - не полный, всего 29 дней из 31.

За сентябрь 2018 запись есть только за 1 день.

Учтем это при определении количества дней в месяце.

In [ ]:
def days_in_month(row):
 if row['order_purchase_year'] == '2016' and row['order_purchase_month'] == 9:
     val = 2
 elif row['order_purchase_year'] == '2016' and row['order_purchase_month'] == 10:
     val = 8
 elif row['order_purchase_year'] == '2016' and row['order_purchase_month'] == 12:
     val = 1
 elif row['order_purchase_year'] == '2017' and row['order_purchase_month'] == 1:
     val = 27
 elif row['order_purchase_year'] == '2018' and row['order_purchase_month'] == 8:
     val = 29
 elif row['order_purchase_year'] == '2018' and row['order_purchase_month'] == 9:
     val = 1
 elif row['order_purchase_month'] in (1,3,5,7,8,10,12):
     val = 31
 elif row['order_purchase_month'] in (4,6,9,11):
     val = 30
 elif row['order_purchase_month'] == 2:
     val = 28
 else:
     val = 0 
 return val

In [ ]:
purch2items['days_in_month'] = purch2items.apply(lambda x: days_in_month(x), axis=1)

Расчет количества недель в месяце:

In [ ]:
purch2items['weeks_in_month'] = purch2items["days_in_month"] / 7

In [ ]:
purch2items = purch2items.groupby(['order_purchase_month','customer_unique_id','weeks_in_month']) \
                         .agg({'order_id':'nunique'}).reset_index() \
                         .rename(columns={"order_id": "orders_count"})

In [ ]:
purch2items['mean_cnt_orders'] = purch2items.orders_count / purch2items.weeks_in_month

In [ ]:
purch2items.head()

order_purchase_month                customer_unique_id  weeks_in_month  \
0                     1  000d460961d6dbfa3ec6c9f5805769e1        4.428571   
1                     1  00115fc7123b5310cf6d3a3aa932699e        4.428571   
2                     1  0014a5a58da615f7b01a4f5e194bf5ea        4.428571   
3                     1  0028cfeb9dfcd3628a56d5b6400deef5        4.428571   
4                     1  002d3bd901608f67c3fc11eaaa842b13        4.428571   

   orders_count  mean_cnt_orders  
0             1         0.225806  
1             1         0.225806  
2             1         0.225806  
3             1         0.225806  
4             1         0.225806

## 5.1. Выполните когортный анализ пользователей.

## 5.2. В период с января по декабрь выявите когорту с самым высоким retention на 3-й месяц. 

In [ ]:
purchases.head()

order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp   order_approved_at  \
0    delivered      2017-10-02 10:56:33 2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37 2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49 2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06 2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39 2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0                    2017-10-18  
1                    2018-08-13  
2                    2018-09-04  
3                    2017-12-15  
4                    2018-02-26

Соединим необходимые данные:

In [ ]:
purchases4cohort = purchases \
        .merge(order_items, how = 'inner', on = 'order_id') \
        .merge(customers, how = 'inner', on = 'customer_id') \
        [['order_approved_at', 'customer_unique_id']]

У нас есть данные о заказах с сентября 2016 по сентябрь 2018:

In [ ]:
purchases4cohort['order_approved_at'].agg(['min','max'])

min   2016-09-15 12:16:38
max   2018-09-03 17:40:06
Name: order_approved_at, dtype: datetime64[ns]

Дата покупки - нужен столбец с первым днем месяца (нужно посчитать транзакции помесячно)

In [ ]:
purchases4cohort['order_approved_month'] = purchases4cohort['order_approved_at'].astype('datetime64[M]')

Но по условию задания нам нужно провести анализ только с января по декабрь. <br>
Делаю вывод, что нужны данные с января 2017 по декабрь 2017 включительно.

Оставим только необходимые месяцы:

In [ ]:
purchases4cohort = purchases4cohort \
                   .query('order_approved_month >= "2017-01-01" & order_approved_month < "2018-01-01"')

In [ ]:
purchases4cohort

order_approved_at                customer_unique_id  \
0      2017-10-02 11:07:15  7c396fd4830fd04220f754e42b4e5bff   
3      2017-11-18 19:45:59  7c142cf63193a1473d2e66489a9ae977   
5      2017-07-09 22:10:13  80bb27c7c16e8f973207a5086ab329e2   
6      2017-04-13 13:25:17  36edbb3fb164b1f16485364b6fb04c73   
7      2017-05-16 13:22:11  932afa1e708222e5821dac9cd5db4cae   
...                    ...                               ...   
112077 2017-10-04 20:07:14  c716cf2b5b86fb24257cffe9e7969df8   
112078 2017-01-27 01:05:25  e03dbdf5e56c96b106d8115ac336f47f   
112079 2017-02-23 09:15:11  831ce3f1bacbd424fc4e38fbd4d66d29   
112080 2017-03-09 09:54:05  6359f309b166b0196dbf7ad2ac62bb5a   
112082 2017-08-27 15:04:16  737520a9aad80b3fbbdad19b66b37b30   

       order_approved_month  
0                2017-10-01  
3                2017-11-01  
5                2017-07-01  
6                2017-04-01  
7                2017-05-01  
...                     ...  
112077           2017-10-01  
112078           2017-01-01  
112079           2017-02-01  
112080           2017-03-01  
112082           2017-08-01  

[50541 rows x 3 columns]

In [ ]:
purchases4cohort['first_order_month'] = purchases4cohort.groupby('customer_unique_id')['order_approved_month'] \
                                                        .transform('min') 

In [ ]:
purchases4cohort = purchases4cohort.rename(columns = {'first_order_month' : 'cohort', 'order_approved_month' : 'cohort_period'})

In [ ]:
purchases4cohort = purchases4cohort.groupby(['cohort','cohort_period'], as_index=False) \
                                   .agg({'customer_unique_id':'nunique'})

In [ ]:
#напишу функцию для определения разницы в месяцах
def month_diff(x, y):
 end = x. dt.to_period('M'). view (dtype='int64')
 start = y. dt.to_period('M'). view (dtype='int64')
 return end-start

In [ ]:
purchases4cohort['month_number'] = month_diff(purchases4cohort.cohort_period,purchases4cohort.cohort)

In [ ]:
purchases4cohort

cohort cohort_period  customer_unique_id  month_number
0  2017-01-01    2017-01-01                 717             0
1  2017-01-01    2017-02-01                   4             1
2  2017-01-01    2017-03-01                   2             2
3  2017-01-01    2017-04-01                   1             3
4  2017-01-01    2017-05-01                   3             4
..        ...           ...                 ...           ...
72 2017-10-01    2017-11-01                  30             1
73 2017-10-01    2017-12-01                  13             2
74 2017-11-01    2017-11-01                7054             0
75 2017-11-01    2017-12-01                  39             1
76 2017-12-01    2017-12-01                5591             0

[77 rows x 4 columns]

In [ ]:
purchases4cohort_pivot = purchases4cohort \
 .pivot_table(index='cohort', columns='month_number',values='customer_unique_id', aggfunc='sum')

In [ ]:
purchases4cohort_pivot

month_number      0     1     2     3     4     5     6    7    8    9    10  \
cohort                                                                         
2017-01-01     717.0   4.0   2.0   1.0   3.0   1.0   3.0  1.0  1.0  NaN  2.0   
2017-02-01    1674.0   4.0   4.0   3.0   7.0   2.0   4.0  3.0  1.0  3.0  2.0   
2017-03-01    2580.0  13.0   9.0   9.0   9.0   4.0   4.0  8.0  8.0  3.0  NaN   
2017-04-01    2302.0  14.0   5.0   4.0   6.0   6.0   8.0  6.0  8.0  NaN  NaN   
2017-05-01    3543.0  19.0  17.0  14.0  11.0  12.0  15.0  5.0  NaN  NaN  NaN   
2017-06-01    3110.0  14.0  11.0  12.0   8.0  11.0  11.0  NaN  NaN  NaN  NaN   
2017-07-01    3777.0  20.0  12.0   7.0  13.0   7.0   NaN  NaN  NaN  NaN  NaN   
2017-08-01    4154.0  28.0  15.0  11.0  16.0   NaN   NaN  NaN  NaN  NaN  NaN   
2017-09-01    4098.0  27.0  20.0  14.0   NaN   NaN   NaN  NaN  NaN  NaN  NaN   
2017-10-01    4360.0  30.0  13.0   NaN   NaN   NaN   NaN  NaN  NaN  NaN  NaN   
2017-11-01    7054.0  39.0   NaN   NaN   NaN   NaN   NaN  NaN  NaN  NaN  NaN   
2017-12-01    5591.0   NaN   NaN   NaN   NaN   NaN   NaN  NaN  NaN  NaN  NaN   

month_number   11  
cohort             
2017-01-01    2.0  
2017-02-01    NaN  
2017-03-01    NaN  
2017-04-01    NaN  
2017-05-01    NaN  
2017-06-01    NaN  
2017-07-01    NaN  
2017-08-01    NaN  
2017-09-01    NaN  
2017-10-01    NaN  
2017-11-01    NaN  
2017-12-01    NaN

In [ ]:
ur_style = (purchases4cohort_pivot
            .style
            .set_caption('User retention by cohort')  # добавляем подпись
            .background_gradient(cmap='viridis')  # раскрашиваем ячейки по столбцам
            .highlight_null('white')  # делаем белый фон для значений NaN
            .format("{:.2f}", na_rep=""))  # числа форматируем как проценты, NaN заменяем на пустоту
ur_style

Когорта с самым высоким retention на 3-й месяц (нужно смотреть month_number = 2) - 2017-05-01

## 6. Используя python, построй RFM-сегментацию пользователей, чтобы качественно оценить свою аудиторию.

Часто для качественного анализа аудитории используют подходы, основанные на сегментации.  

В кластеризации можешь выбрать следующие метрики: 

R - время от последней покупки пользователя до текущей даты, 

F - суммарное количество покупок у пользователя за всё время, 

M - сумма покупок за всё время. 

Подробно опиши, как ты создавал кластеры. Для каждого RFM-сегмента построй границы метрик recency, frequency и monetary для интерпретации этих кластеров. 

Пример такого описания: RFM-сегмент 132 (recency=1, frequency=3, monetary=2) имеет границы метрик recency от 130 до 500 дней, frequency от 2 до 5 заказов в неделю, monetary от 1780 до 3560 рублей в неделю. 

Описание подхода можно найти тут.

Для данного пункта нам понадобится датасет с покупками purchases + нужны данные по составу каждого заказа - olist_order_items_dataset

In [ ]:
purch2items_6 = purchases \
        .merge(order_items, how = 'inner', on = 'order_id') \
        .merge(customers, how = 'inner', on = 'customer_id') \
        [['order_id', 'order_approved_at', 'customer_unique_id', 'order_purchase_timestamp', 'order_item_id', 'product_id', 'price']]

Сейчас 2024 год, и если мы будем брать текущую дату как точку отсчета, чтобы рассчитать показатель **recency** - выйдет некорректно. Поэтому от последней даты покупки отсчитаем 1 день, и будем от этой даты отсчитывать давность покупок. <br>

Буду использовать дату *order_purchase_timestamp*.

In [ ]:
snapshot_date = purch2items_6['order_purchase_timestamp'].max() + timedelta(days=1)
print(snapshot_date)

2018-09-04 09:06:57


Теперь можно рассчитать показатели. <br>
<br>
**Recency** - это количество дней с последней покупки до даты *snapshot_date*<br>
**Frequency** - это количество покупок, в нашем случае - это количество заказов (нужно считать уникальные *order_id*).<br>
**Monetary** - это сумма по столбцу *price*.<br>
   

In [ ]:
rfm_processing = purch2items_6.groupby(['customer_unique_id']).agg({ \
             'order_purchase_timestamp': lambda x: (snapshot_date - x.max()).days, \
             'order_id': 'nunique',  \
             'price': 'sum' })

rfm_processing.rename(columns={'order_purchase_timestamp':'recency', \
                               'order_id':'frequency', \
                               'price':'monetary'}, inplace=True)

In [ ]:
print(rfm_processing.head())
print('В датафрейме {} строк и {} столбцов.' \
     .format(rfm_processing.shape[0], rfm_processing.shape[1]))

                                  recency  frequency  monetary
customer_unique_id                                            
0000366f3b9a7992bf8c76cfdf3221e2      116          1    129.90
0000b849f77a49e4a4ce2b2a4ca5be3f      119          1     18.90
0000f46a3911fa3c0805444483337064      542          1     69.00
0000f6ccb0745a6a4b88665a16c9f078      326          1     25.99
0004aac84e0df4da2b147fca70cf8255      293          1    180.00
В датафрейме 94970 строк и 3 столбцов.


Итак, у нас есть 94970 пользователей, которых нужно сегментировать с помощью методов RFM-анализа. <br>

Нужно разбить показатели **recency**, **frequency** и **monetary** на бакеты. 
По сути, каждому значению мы присвоим балл, который показывает нам, насколько "хорошее" это для нас значение.
Будем оценивать показатели от 1 до 5. <br>
<br>
**recency** - как давно что-то покупал? Чем более недавно, тем лучше<br>
**frequency** - как часто покупал? Чем чаще, тем лучше<br>
**monetary** - как много денег у нас потратил? Чем больше, тем лучше<br>

Воспользуемся методом **qcut** <br>

In [ ]:
pd.qcut(rfm_processing['recency'],5)

customer_unique_id
0000366f3b9a7992bf8c76cfdf3221e2     (98.0, 182.0]
0000b849f77a49e4a4ce2b2a4ca5be3f     (98.0, 182.0]
0000f46a3911fa3c0805444483337064    (388.0, 729.0]
0000f6ccb0745a6a4b88665a16c9f078    (273.0, 388.0]
0004aac84e0df4da2b147fca70cf8255    (273.0, 388.0]
                                         ...      
fffcf5a5ff07b0908bd4e2dbc735a684    (388.0, 729.0]
fffea47cd6d3cc0a88bd621562a9d061    (182.0, 273.0]
ffff371b4d645b6ecea244b27531430a    (388.0, 729.0]
ffff5962728ec6157033ef9805bacc48     (98.0, 182.0]
ffffd2657e2aad2907e67c3e9daecbeb    (388.0, 729.0]
Name: recency, Length: 94970, dtype: category
Categories (5, interval[float64]): [(0.999, 98.0] < (98.0, 182.0] < (182.0, 273.0] < (273.0, 388.0] < (388.0, 729.0]]

Показатель **recency** будем оценивать так: <br>
+ (0.999, 98.0] 5 баллов, так как это самые недавние по времени заказы <br>
+ (98.0, 182.0] 4 балла <br>
+ (182.0, 273.0] 3 балла <br>
+ (273.0, 388.0] 2 балла <br>
+ (388.0, 729.0] 1 балл, это самые давние покупки <br>

Проставляем баллы <br>

In [ ]:
rfm_processing['r_score'] = 6 - (pd.qcut(rfm_processing['recency'],5,labels=False) + 1)

Убедимся, что наша задача выполнена:

In [ ]:
rfm_processing.groupby('r_score')['recency'].agg(['mean', 'count'])

mean  count
r_score                   
1        479.345760  18964
2        321.860530  18936
3        225.598885  19012
4        141.084913  18937
5         50.664714  19121

Действительно разделили показатели на 5 приблизительно равных групп. <br>
<br>

И учли тот факт, что в случае **recency** чем больше показатель, тем ниже балл, потому что большое значение **recency** означает, что покупатель был у нас очень давно.

С показателем **frequency** этот номер не пройдет. <br>
Слишком часто повторяются одинаковые значения, и мы не сможем разбить их на группы с помощью *qcut*. <br>
Продемонстрируем это:

In [ ]:
rfm_processing['frequency'].value_counts()

1     92084
2      2650
3       188
4        29
5         9
6         5
7         3
16        1
9         1
Name: frequency, dtype: int64

## И здесь придется применять индивидуальный подход для формирования бакетов.<br>
Разобъем эти значения на группы вручную.<br>
<br>
1 балл я поставлю тем пользователям, которые совершили только одну покупку. <br>
2 балла я поставлю тем пользователям, у которых 2 покупки.<br>
3 балла я поставлю пользователям, которые сделали 3-4 покупки.<br>
4 балла получат пользователи, у которых от 5-6 покупок.<br>
5 баллов получат пользователи, у которых больше или равно 7 покупок.<br>


In [ ]:
def get_f_score(x):
    if x == 1:
        return 1
    if x == 2:
        return 2
    if (x >= 3) & (x <= 4):
        return 3
    if (x >= 5) & (x <= 6):
        return 4
    if x >= 7:
        return 5

In [ ]:
rfm_processing['f_score'] = rfm_processing['frequency'].apply(get_f_score)

In [ ]:
rfm_processing.groupby('f_score')['frequency'].agg(['mean','count'])

mean  count
f_score                 
1        1.000000  92084
2        2.000000   2650
3        3.133641    217
4        5.357143     14
5        9.200000      5

Теперь приступим к показателю **monetary**. <br>
Также будем использовать метод **qcut**.

In [ ]:
pd.qcut(rfm_processing['monetary'],5)

customer_unique_id
0000366f3b9a7992bf8c76cfdf3221e2      (109.9, 179.9]
0000b849f77a49e4a4ce2b2a4ca5be3f       (0.849, 39.9]
0000f46a3911fa3c0805444483337064        (39.9, 69.9]
0000f6ccb0745a6a4b88665a16c9f078       (0.849, 39.9]
0004aac84e0df4da2b147fca70cf8255    (179.9, 13440.0]
                                          ...       
fffcf5a5ff07b0908bd4e2dbc735a684    (179.9, 13440.0]
fffea47cd6d3cc0a88bd621562a9d061        (39.9, 69.9]
ffff371b4d645b6ecea244b27531430a       (69.9, 109.9]
ffff5962728ec6157033ef9805bacc48      (109.9, 179.9]
ffffd2657e2aad2907e67c3e9daecbeb        (39.9, 69.9]
Name: monetary, Length: 94970, dtype: category
Categories (5, interval[float64]): [(0.849, 39.9] < (39.9, 69.9] < (69.9, 109.9] < (109.9, 179.9] < (179.9, 13440.0]]

Оценивать **monetary** будем так:
+ (0.849, 39.9] - 1 балл
+ (39.9, 69.9] - 2 балла
+ (69.9, 109.9] - 3 балла
+ (109.9, 179.9] - 4 балла
+ (179.9, 13440.0] - 5 баллов

Проставляем баллы:

In [ ]:
rfm_processing['m_score'] = pd.qcut(rfm_processing['monetary'],5,labels=False) + 1

Проверим:

In [ ]:
rfm_processing.groupby('m_score')['monetary'].agg(['mean','count'])

mean  count
m_score                   
1         26.080639  19638
2         55.143896  19732
3         90.139484  17720
4        140.433940  18975
5        403.620931  18905

Можно рассчитать итоговую оценку:

In [ ]:
rfm_processing['rfm_score'] = rfm_processing['r_score'].map(str) \
                            + rfm_processing['f_score'].map(str) \
                            + rfm_processing['m_score'].map(str)

In [ ]:
rfm_processing.sort_values('rfm_score', ascending=False)

recency  frequency  monetary  r_score  \
customer_unique_id                                                        
6469f99c1f9dfae7733b25662e7f1782       68          7    664.20        5   
8d50f5eadf50201ccdcedfb9e2ac8455       14         16    729.62        5   
ca77025e7201e3b30c44b472ff346268       94          7    806.61        5   
dc813062e0fc23409cd255f7f53c7074       12          6    687.76        5   
394ac4de8f3acb14253c177f0e15bc58       19          5    633.16        5   
...                                   ...        ...       ...      ...   
634bba2731b949e84282ef83f2b28490      447          1     39.90        1   
635efb83a8893712f25acc9bb44abc48      513          1     11.99        1   
6360a310e76374869840970850ea3135      557          1     24.90        1   
637a39768880ddca5882b455322ff41f      541          1     14.90        1   
74bba753d0dfd62b4c32ee45cab7c812      490          1     34.90        1   

                                  f_score  m_score rfm_score  
customer_unique_id                                            
6469f99c1f9dfae7733b25662e7f1782        5        5       555  
8d50f5eadf50201ccdcedfb9e2ac8455        5        5       555  
ca77025e7201e3b30c44b472ff346268        5        5       555  
dc813062e0fc23409cd255f7f53c7074        4        5       545  
394ac4de8f3acb14253c177f0e15bc58        4        5       545  
...                                   ...      ...       ...  
634bba2731b949e84282ef83f2b28490        1        1       111  
635efb83a8893712f25acc9bb44abc48        1        1       111  
6360a310e76374869840970850ea3135        1        1       111  
637a39768880ddca5882b455322ff41f        1        1       111  
74bba753d0dfd62b4c32ee45cab7c812        1        1       111  

[94970 rows x 7 columns]

Получили список пользователей, упорядоченный по *rfm_score*, самые "лучшие" покупатели - в начале списка. <br>
А в конце списка - "оттекающие" покупатели. Они делали покупки очень давно, покупок было очень мало, и потратили они немного.

Теперь разобъем пользователей на сегменты. <br>

+ **Новички** - в эту категорию включим покупателей, которые сделали 1 покупку в течение последних 98 дней
+ **Дрейфующие** - сюда попадут покупатели, которые делали 1 заказ в период от 98 до 273 дней или от 2 до 4 заказов от 182 до 273 дней назад
+ **Спящие** - от 1 до 4 заказов, ранее чем 273 дня назад
+ **Перспективные** - от 2 до 4 заказов, 182 и меньше дней назад
+ **Лояльные** - от 5 и больше покупок, ранее чем 182 дня назад
+ **В зоне риска** - от 5 и больше покупок, от 182 до 388 дней назад
+ **В зоне потери** - от 5 и больше покупок, больше чем 273 дня назад

In [ ]:
segt_map = {
    r'51': 'новички',
    r'[3-4]1': 'дрейфующие',
    r'3[2-3]': 'дрейфующие',
    r'[1-2][1-3]': 'спящие',
    r'[4-5][2-3]': 'перспективные',
    r'[4-5][4-5]': 'лояльные',
    r'[2-3][4-5]': 'в зоне риска',
    r'[1][4-5]': 'в зоне потери'}    

rfm_processing['Segment'] = rfm_processing['r_score'].map(str) + rfm_processing['f_score'].map(str)
rfm_processing['Segment'] = rfm_processing['Segment'].replace(segt_map, regex=True)
rfm_processing.head()

recency  frequency  monetary  r_score  \
customer_unique_id                                                        
0000366f3b9a7992bf8c76cfdf3221e2      116          1    129.90        4   
0000b849f77a49e4a4ce2b2a4ca5be3f      119          1     18.90        4   
0000f46a3911fa3c0805444483337064      542          1     69.00        1   
0000f6ccb0745a6a4b88665a16c9f078      326          1     25.99        2   
0004aac84e0df4da2b147fca70cf8255      293          1    180.00        2   

                                  f_score  m_score rfm_score     Segment  
customer_unique_id                                                        
0000366f3b9a7992bf8c76cfdf3221e2        1        4       414  дрейфующие  
0000b849f77a49e4a4ce2b2a4ca5be3f        1        1       411  дрейфующие  
0000f46a3911fa3c0805444483337064        1        2       112      спящие  
0000f6ccb0745a6a4b88665a16c9f078        1        1       211      спящие  
0004aac84e0df4da2b147fca70cf8255        1        5       215      спящие

In [ ]:
rfm_level_agg = rfm_processing.groupby('Segment').agg({
    'recency' : ['mean'],
    'frequency' : ['mean'],
    'monetary' : ['mean','count']}).round(1)

Получаем вот такое распределение пользователей по сегментам:

In [ ]:
rfm_level_agg.columns = rfm_level_agg.columns.droplevel()
rfm_level_agg.columns = ['Ср_дней_с_посл_покупки','Ср_кол-во_покупок','Ср_выручка','Кол-во_чел']
rfm_level_agg.sort_values(by='Кол-во_чел', ascending=False)

Ср_дней_с_посл_покупки  Ср_кол-во_покупок  Ср_выручка  \
Segment                                                                
спящие                          400.7                1.0       142.9   
дрейфующие                      184.1                1.0       138.2   
новички                          50.6                1.0       139.3   
перспективные                    93.9                2.1       265.3   
лояльные                         99.8                6.3       645.5   
в зоне риска                    210.0                6.8       783.2   
в зоне потери                   606.0                6.0        58.4   

               Кол-во_чел  
Segment                    
спящие              37899  
дрейфующие          37352  
новички             18473  
перспективные        1227  
лояльные               14  
в зоне риска            4  
в зоне потери           1